ttnn/cpp/ttnn/tensor/xtensor/conversion_utils.cpp

```diff
@@ -0,0 +1,163 @@
+    auto tiled_buffer = tt::tt_metal::host_buffer::get_as<InternalT>(cpu_tensor);
+    auto untiled_shape = cpu_tensor.get_logical_shape();
+    auto tiled_shape = cpu_tensor.get_padded_shape();
+
+    // Calculate total size of the untiled tensor
+    size_t total_size = untiled_shape.volume();
+
+    std::vector<T> untiled_data(total_size);
+
+    auto compute_flat_index = [](const std::vector<uint32_t>& indices, ttnn::SimpleShape& shape) -> uint32_t {
+        uint32_t flat_index = 0;
+        uint32_t multiplier = 1;
+        for (int i = (int)indices.size() - 1; i >= 0; --i) {
+            flat_index += indices[i] * multiplier;
+            multiplier *= shape[i];
+        }
+        return flat_index;
+    };
+
+    std::vector<uint32_t> indices(tiled_shape.rank(), 0);
+
+    for (size_t idx = 0; idx < total_size; ++idx) {
+        uint32_t untiled_index = compute_flat_index(indices, untiled_shape);
+        uint32_t tiled_index = compute_flat_index(indices, tiled_shape);
+        if constexpr (std::is_same_v<InternalT, bfloat16>) {
+            untiled_data[untiled_index] = tiled_buffer[tiled_index].to_float();
+        } else {
+            untiled_data[untiled_index] = tiled_buffer[tiled_index];
+        }
+
+        for (int dim = (int)tiled_shape.rank() - 1; dim >= 0; --dim) {
+            if (++indices[dim] < untiled_shape[dim]) {
+                break;
+            }
+            indices[dim] = 0;
+        }
+    }
+
+    return untiled_data;
+}
+
+}  // namespace
+
+template <>
+Tensor from_vector<float, DataType::BFLOAT16>(const std::vector<float>& buffer, const ttnn::Shape& shape) {
+    const DataType data_type = DataType::BFLOAT16;
+    auto logical_shape = shape.logical_shape();
+    size_t volume = logical_shape.volume();
+    TT_FATAL(
+        buffer.size() == volume, "Current buffer size is {} different from shape volume {}", buffer.size(), volume);
+    auto owned_buffer = create_owned_buffer(buffer, data_type);
+    return Tensor(OwnedStorage{owned_buffer}, logical_shape, data_type, Layout::ROW_MAJOR);
+}
+
+// Workaround implementation due to issue with tilize for float32
+// it is expected that tilize will be fixed in the after next tt-metal main update
+template <>
+Tensor from_vector<float, DataType::FLOAT32>(const std::vector<float>& buffer, const ttnn::Shape& shape) {
+    auto tensor = from_vector<float, DataType::BFLOAT16>(buffer, shape);
+    return ttnn::typecast(tensor, DataType::FLOAT32);
+}
+
+template <>
+Tensor from_vector<uint32_t, DataType::UINT32>(const std::vector<uint32_t>& buffer, const ttnn::Shape& shape) {
+    MemoryConfig output_mem_config{};
+    auto logical_shape = shape.logical_shape();
+    auto volume = logical_shape.volume();
+    TT_FATAL(
+        buffer.size() == volume, "Current buffer size is {} different from shape volume {}", buffer.size(), volume);
+    return create_owned_tensor(buffer, logical_shape, DataType::UINT32, Layout::ROW_MAJOR);
```

Comment:
Was your plan to extend this later for other Layout types or exclusively `Layout::RowMajor`?

Reply:
In this particular place, at least for now, I don't think so, as we expect to run this on the host.